File: core-ui/runtime/src/toggleaction.js
// ToggleAction runtime — the three-state cousin of optimisticaction.
// optimisticaction commits once and stays committed; ToggleAction
// supports two additional UX patterns that the existing widget
// explicitly blocked (V3 #10):
//
//   1. data-fui-toggle-group="<key>"  — buttons sharing the same group
//      key form a mutex. Committing one auto-revokes any sibling that
//      was previously committed (no extra RPC, the server is the
//      source of truth; the UI flips optimistically and a subsequent
//      navigate refreshes from server state).
//
//   2. data-fui-toggle-allow-untoggle="true" — clicking an already-
//      committed button reverts it to idle. POSTs the untoggle
//      endpoint (data-fui-toggle-untoggle-endpoint) if set, otherwise
//      just flips locally. Without this, the button is sticky once
//      committed, matching optimisticaction's behaviour.
//
// Each <button data-fui-comp="ui-toggle-action"> declares its commit
// endpoint via data-fui-toggle-endpoint and (optionally) method via
// data-fui-toggle-method (defaults POST). SSR ships the initial state
// in data-state="idle|committed". The widget renders two child spans
// (data-fui-toggle-idle / data-fui-toggle-committed) — the runtime
// shows/hides via the [hidden] attribute, no inline CSS required.
(() => {
  'use strict';

  const setState = (btn, state) => {
    btn.setAttribute('data-state', state);
    const idle = btn.querySelector('[data-fui-toggle-idle]');
    const committed = btn.querySelector('[data-fui-toggle-committed]');
    if (!idle || !committed) return;
    if (state === 'committed' || state === 'pending') {
      idle.setAttribute('hidden', '');
      committed.removeAttribute('hidden');
    } else {
      idle.removeAttribute('hidden');
      committed.setAttribute('hidden', '');
    }
    if (state === 'pending') {
      btn.setAttribute('aria-busy', 'true');
      btn.disabled = true;
    } else {
      btn.removeAttribute('aria-busy');
      btn.disabled = false;
    }
    // aria-pressed mirrors the binary toggle state for AT users — the
    // "committed" state IS the pressed-toggle convention.
    btn.setAttribute('aria-pressed', state === 'committed' ? 'true' : 'false');
  };

  const csrfHeaders = () => {
    const headers = { Accept: 'application/json' };
    const meta = document.querySelector('meta[name="csrf-token"]');
    if (meta) {
      const tok = meta.getAttribute('content');
      if (tok) headers['X-CSRF-Token'] = tok;
    }
    return headers;
  };

  const fireAndForget = (url, method) => {
    if (!url) return Promise.resolve(true);
    return fetch(url, {
      method: (method || 'POST').toUpperCase(),
      credentials: 'same-origin',
      headers: csrfHeaders(),
    }).then((res) => res.ok, () => false);
  };

  const revokeGroupSiblings = (group, except) => {
    if (!group) return;
    // CSS.escape: a quote/backslash in the group key must not blow up
    // the selector (an unescaped throw here would strand the clicked
    // button in its pending/disabled state).
    const sel = `[data-fui-comp="ui-toggle-action"][data-fui-toggle-group="${CSS.escape(group)}"]`;
    for (const other of document.querySelectorAll(sel)) {
      if (other === except) continue;
      if (other.getAttribute('data-state') === 'committed') {
        setState(other, 'idle');
      }
    }
  };

  const setupOne = (btn) => {
    if (btn.__fuiToggleBound) return;
    btn.__fuiToggleBound = true;
    btn.addEventListener('click', (ev) => {
      ev.preventDefault();
      if (btn.getAttribute('data-state') === 'pending') return;
      const state = btn.getAttribute('data-state') || 'idle';
      const group = btn.getAttribute('data-fui-toggle-group') || '';
      const allowUntoggle =
        (btn.getAttribute('data-fui-toggle-allow-untoggle') || '').toLowerCase() === 'true';
      const commitURL = btn.getAttribute('data-fui-toggle-endpoint');
      const untoggleURL = btn.getAttribute('data-fui-toggle-untoggle-endpoint');
      const method = btn.getAttribute('data-fui-toggle-method') || 'POST';

      if (state === 'committed') {
        if (!allowUntoggle) return; // sticky — matches optimisticaction default
        setState(btn, 'pending');
        btn.dispatchEvent(new CustomEvent('toggle-action:untoggle', { bubbles: true }));
        fireAndForget(untoggleURL, method).then((ok) => {
          setState(btn, ok ? 'idle' : 'committed');
        });
        return;
      }

      // state === 'idle' — commit, and revoke any sibling in the group.
      setState(btn, 'pending');
      revokeGroupSiblings(group, btn);
      btn.dispatchEvent(new CustomEvent('toggle-action:commit', { bubbles: true }));
      fireAndForget(commitURL, method).then((ok) => {
        setState(btn, ok ? 'committed' : 'idle');
      });
    });
  };

  const scan = (root) => {
    const scope = root && root.querySelectorAll ? root : document;
    for (const btn of scope.querySelectorAll('[data-fui-comp="ui-toggle-action"]')) {
      setupOne(btn);
    }
  };

  requestAnimationFrame(() => scan(document));
  document.addEventListener('gofastr:navigate', () => {
    requestAnimationFrame(() => scan(document));
  });

  window.__gofastr = window.__gofastr || {};
  window.__gofastr._moduleScanners = window.__gofastr._moduleScanners || {};
  window.__gofastr._moduleScanners.toggleaction = scan;
  window.__gofastr.toggleaction = { rescan: scan };
})();

File: framework/ui/toggleaction.go
package ui

import (
	"context"
	"strings"

	"github.com/DonaldMurillo/gofastr/core-ui/html"
	"github.com/DonaldMurillo/gofastr/core-ui/registry"
	"github.com/DonaldMurillo/gofastr/core-ui/style"
	"github.com/DonaldMurillo/gofastr/core/render"
	"github.com/DonaldMurillo/gofastr/framework/headless"
)

// The behaviour is the headless module's: the primitive's
// data-hui-action* hooks bind through the kernel's action primitive,
// and the retired toggleaction adapter module (its .js and its
// registration) is deleted with this move.

// ─── ToggleAction ───────────────────────────────────────────────────
//
// The three-state cousin of OptimisticAction. OptimisticAction commits
// once and stays committed; ToggleAction cycles idle → pending →
// committed and supports two additional patterns:
//
//  1. Mutex groups (Group): buttons sharing the same group key form a
//     mutex: committing one optimistically reverts any sibling that
//     was committed (no extra RPC; the server stays the source of
//     truth and a later navigation refreshes from server state).
//
//  2. Untoggle (AllowUntoggle / UntoggleEndpoint): clicking an
//     already-committed button reverts it to idle. If UntoggleEndpoint
//     is set the runtime POSTs it; otherwise the state just flips
//     locally. Without either, the button is sticky once committed,
//     matching OptimisticAction's behaviour.
//
// Use for "Follow / Following", plan pickers, watch/unwatch: binary
// server-backed state the user flips in place. For one-shot commits
// prefer OptimisticAction; for destructive actions pair with
// ConfirmAction instead.
//
// SSR shape (state ships server-rendered; the runtime only flips it):
//
//	<button data-fui-comp="ui-toggle-action"
//	        data-hui-action="" data-hui-action-endpoint="/follow"
//	        data-hui-action-untoggle="/unfollow"
//	        data-hui-action-group="follows"
//	        data-hui-action-failed="…"
//	        data-state="idle" aria-pressed="false"
//	        class="fui-button fui-button--primary fui-toggle-action">
//	    <span data-hui-action-idle>Follow</span>
//	    <span data-hui-action-done hidden>Following ✓</span>
//	</button>
//
// The headless module binds the data-hui-action* hooks through the
// kernel's action primitive (the retired toggleaction.js is deleted)
// and mirrors the committed state onto aria-pressed.

// ToggleActionConfig configures a ToggleAction button.
type ToggleActionConfig struct {
	// Endpoint is the URL hit when toggling idle → committed. Required.
	Endpoint string

	// Method is "POST" (default), "DELETE", "PATCH", or "PUT". Applies
	// to both the commit and the untoggle request.
	Method string

	// IdleLabel is the button text in the un-committed state. Required.
	IdleLabel string

	// CommittedLabel is shown while committed (the runtime flips to it
	// optimistically on click). Required.
	CommittedLabel string

	// IdleIcon optionally renders alongside IdleLabel.
	IdleIcon render.HTML

	// CommittedIcon optionally renders alongside CommittedLabel.
	CommittedIcon render.HTML

	// Committed sets the SSR initial state. Render true when the
	// server already knows the action is active (user follows, plan
	// selected) so first paint matches server state.
	Committed bool

	// Group, when set, joins this button to a client-side mutex:
	// committing any button with the same Group key reverts the
	// previously-committed sibling. Maps to data-hui-action-group.
	Group string

	// AllowUntoggle lets a click on a committed button revert it to
	// idle. Maps to the data-hui-action-untoggle hook (empty for a
	// local flip, the untoggle endpoint's URL when one is set).
	AllowUntoggle bool

	// UntoggleEndpoint is the URL hit when reverting committed → idle.
	// Setting it implies AllowUntoggle. When empty (with AllowUntoggle
	// true) the revert flips locally with no request.
	UntoggleEndpoint string

	// Variant maps to the standard Button variant ("primary"/""
	// default, "secondary", "danger", "ghost").
	Variant ButtonVariant

	// Size maps to the standard Button size ("" default, "small",
	// "large").
	Size ButtonSize

	ID    string
	Class string

	// ExtraAttrs forwards additional attributes (data-* test hooks,
	// analytics markers) to the button's root <button> element. Keys
	// the component owns are dropped: class and id (use Class / ID),
	// data-fui-* (the toggle runtime wiring), type, data-state, and
	// aria-pressed (mirrored from Committed by the runtime).
	ExtraAttrs html.Attrs

	// FailedText is what the status span announces on a failure.
	// Empty takes the Strings default.
	FailedText string
	// Ctx carries the per-request context used to resolve the
	// failure sentence. When nil, English fallbacks apply.
	Ctx context.Context
	// Disabled is the state at render time.
	Disabled bool
}

// ToggleAction renders the button. The marker fetches this sheet; the
// clicks are bound through the data-hui-action* hooks the primitive
// renders.
func ToggleAction(cfg ToggleActionConfig) render.HTML {
	if cfg.Endpoint == "" {
		panic("ui: ToggleAction requires Endpoint")
	}
	if cfg.IdleLabel == "" {
		panic("ui: ToggleAction requires IdleLabel")
	}
	if cfg.CommittedLabel == "" {
		panic("ui: ToggleAction requires CommittedLabel")
	}
	tv := cfg.Variant
	if tv == "" {
		tv = ButtonPrimary
	}
	checkButtonVariant("ToggleAction", tv)
	checkButtonSize("ToggleAction", cfg.Size)

	parts := headless.Parts{}
	if extra := strings.TrimSpace(buttonClassTokens(tv, cfg.Size) + " fui-toggle-action " + cfg.Class); extra != "" {
		parts.Attrs = headless.PartAttrs{headless.PartRoot: {"class": strings.TrimSpace(extra)}}
	}
	ctx := cfg.Ctx
	if ctx == nil {
		ctx = context.Background()
	}
	return toggleActionStyle.WrapHTML(headless.ToggleAction(headless.ToggleActionProps{
		Endpoint:         cfg.Endpoint,
		Method:           cfg.Method,
		IdleLabel:        cfg.IdleLabel,
		CommittedLabel:   cfg.CommittedLabel,
		IdleIcon:         cfg.IdleIcon,
		DoneIcon:         cfg.CommittedIcon,
		Committed:        cfg.Committed,
		Group:            cfg.Group,
		AllowUntoggle:    cfg.AllowUntoggle,
		UntoggleEndpoint: cfg.UntoggleEndpoint,
		Variant:          string(cfg.Variant),
		Size:             string(cfg.Size),
		Disabled:         cfg.Disabled,
		FailedText:       cfg.FailedText,
		ID:               cfg.ID,
		ExtraAttrs:       headless.Safe(cfg.ExtraAttrs, "type", "data-state", "aria-pressed"),
		Parts:            parts,
		Strings:          StringsFor(ctx),
	}, toggleActionClasses))
}

// toggleActionClasses dresses the primitive's parts in the button
// family's vocabulary.
var toggleActionClasses = headless.Classes{
	headless.PartRoot: "fui-button",
}

var toggleActionStyle = registry.RegisterStyle("ui-toggle-action", func(_ style.Theme) string {
	return `[data-fui-comp="ui-toggle-action"] {
  /* Inherits .fui-button base; override only what the toggle flip needs. */
  position: relative;
  transition: background-color 120ms ease, color 120ms ease;
}
/* Committed state — success tone signals "active". */
[data-fui-comp="ui-toggle-action"][data-state="committed"] {
  background: var(--color-success, #16A34A);
  color: var(--color-primary-fg, #FFFFFF);
  border-color: var(--color-success, #16A34A);
}
/* Pending — same look as committed (optimistic) plus a busy cursor
   while the RPC is in flight. The runtime also sets aria-busy +
   disabled during this window. */
[data-fui-comp="ui-toggle-action"][data-state="pending"] {
  cursor: progress;
  background: var(--color-success, #16A34A);
  color: var(--color-primary-fg, #FFFFFF);
  border-color: var(--color-success, #16A34A);
}
`
})
